lib/kubernetes-deploy/kubernetes_resource.rb

```diff
@@ -22,27 +22,14 @@ class KubernetesResource
       If you have reason to believe it will succeed, retry the deploy to continue to monitor the rollout.
       MSG
 
-    def self.for_type(type:, name:, namespace:, context:, file:, logger:)
-      subclass = case type
-      when 'cloudsql' then Cloudsql
-      when 'configmap' then ConfigMap
-      when 'deployment' then Deployment
-      when 'pod' then Pod
-      when 'redis' then Redis
-      when 'bugsnag' then Bugsnag
-      when 'ingress' then Ingress
-      when 'persistentvolumeclaim' then PersistentVolumeClaim
-      when 'service' then Service
-      when 'podtemplate' then PodTemplate
-      when 'poddisruptionbudget' then PodDisruptionBudget
-      end
-
-      opts = { name: name, namespace: namespace, context: context, file: file, logger: logger }
-      if subclass
-        subclass.new(**opts)
+    def self.build(namespace:, context:, template:, logger:)
+      opts = { namespace: namespace, context: context, template: template, logger: logger }
+      if KubernetesDeploy.const_defined?(template["kind"])
+        klass = KubernetesDeploy.const_get(template["kind"])
+        klass.new(**opts)
       else
         inst = new(**opts)
-        inst.tap { |r| r.type = type }
+        inst.tap { |r| r.type = template["kind"] }
```

Comment:
I meant to say, I rather the setter function be exposed than using tap; I rarely find `tap` to be necessary.